docs/src/design/chiplets/kernel_rom.md

```diff
@@ -1,91 +1,101 @@
+_Note: Unless otherwise stated, these constraints should also be multiplied by chiplets module's selector
+flag $chip\_s_4$ for the kernel ROM chiplet, as is true for all constraints in this chiplet._
 
-The value in the $idx$ column must either stay the same or increase by $1$.
+The $s_{first}$ column must be binary.
 
 > $$
-\Delta idx \cdot (1 - \Delta idx) = 0 \text{ | degree} = 2
+s_{first}^2 - s_{first} = 0 \text{ | degree} = 2
 $$
 
-Finally, if the $idx$ column stays the same then the kernel procedure root must not change. This can be achieved by enforcing the following constraint against each of the four procedure root columns:
+$s_{first}$ must be set in the first row of the trace, indicating the start of a new hash which must be included in the set of kernel procedure roots provided by the verifier.
+This constraint is enforced in the last row of the previous trace, using selector columns from the [chiplets](main.md)
+module.
+The virtual $chip\_s_3$ flag is active in all rows of the previous chiplet.
+In the last row of that chiplet, the selector $s_3$ transitions from 0 to 1.
 
 > $$
-(1 - \Delta idx) \cdot (r_i' - r_i) = 0 \text{ | degree} = 2
+chip\_s_3 \cdot s_3' \cdot (1 - s_{first}') = 0 \text{ | degree} = \deg(chip\_s_3) + 2
 $$
 
-These constraints on $idx$ should not be applied to the very last row of the kernel ROM's execution trace, since we do not want to enforce a value that would conflict with the first row of a subsequent chiplet (or padding). Therefore we can create a special virtual flag for this constraint using the $chip\_s_3$ selector column from the [chiplets](main.md) module that selects for the kernel ROM chiplet.
+_Note that this selector need not be multiplied by the kernel ROM chiplet flag $chip\_s_4$, since it is only active when the previous chiplet is active._
 
-The modified constraints which should be applied are the following:
+The hash of the procedure repeats in the next row whenever the next row is not the first of a new procedure, i.e., when $s_{first}' = 0$.
+To ensure the constraint is disabled in the last row, we multiply it by the chiplet selector $s_4$ for the next row,
+since it transitions from 0 to 1 at this point, indicating the start of
 
->$$
-(1 - chip\_s_3') \cdot \Delta idx \cdot (1 - \Delta idx) = 0 \text{ | degree} = 3
+> $$
+s_4' \cdot (1 - s_{first}') \cdot (r_i' - r_i) = 0 \text{ | degree} = 3
 $$
 
->$$
-(1 - chip\_s_3') \cdot (1 - \Delta idx) \cdot (r_i' - r_i) = 0 \text{ | degree} = 3
-$$
+### Kernel procedure table constraints
 
-_Note: these constraints should also be multiplied by chiplets module's selector flag for the kernel ROM chiplet, as is true for all constraints in this chiplet._
+This kernel procedure table keeps track of all *unique* kernel function roots. The values in this table will be updated
+only when the value in the $s_{first}$ selector is set.
 
-## Chiplets bus constraints
+The chiplet communicates with the chiplet bus via the $b_{chip}$ columns.
```

Comment:
This makes sense but I wounder if we can somehow and succinctly say the following:

1. The virtual table is a set i.e., each entry exists only once, that contains all digests for the kernel against which the program being verified was compiled,
2. The ROM chiplet makes a read-once request for each digest in the kernel
3. The READ and WRITE sets must be equal